api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2872,293 +2877,250 @@ public void remove() {
+        return true;
+      }
 
-            if (this.error != null) {
-              return true;
-            }
+      if (this.prefixIterator.hasNext()) {
+        return true;
+      }
 
-            if (this.itemIterator.hasNext()) {
-              return true;
-            }
+      this.completed = true;
+      return false;
+    }
 
-            if (this.prefixIterator.hasNext()) {
-              return true;
-            }
+    @Override
+    public Result<Item> next() {
+      if (this.completed) {
+        throw new NoSuchElementException();
+      }
 
-            this.completed = true;
-            return false;
-          }
+      if (this.error == null && this.itemIterator == null && this.prefixIterator == null) {
+        populate();
+      }
 
-          @Override
-          public Result<Item> next() {
-            if (this.completed) {
-              throw new NoSuchElementException();
-            }
+      if (this.error == null
+          && !this.itemIterator.hasNext()
+          && !this.prefixIterator.hasNext()
+          && this.listBucketResult.isTruncated()) {
+        populate();
+      }
 
-            if (this.error == null && this.itemIterator == null && this.prefixIterator == null) {
-              populate();
-            }
+      if (this.error != null) {
+        this.completed = true;
+        return this.error;
+      }
 
-            if (this.error == null
-                && !this.itemIterator.hasNext()
-                && !this.prefixIterator.hasNext()
-                && this.listBucketResult.isTruncated()) {
-              populate();
-            }
+      if (this.itemIterator.hasNext()) {
+        Item item = this.itemIterator.next();
+        this.lastObjectName = item.objectName();
+        return new Result<>(item);
+      }
+      if (this.prefixIterator.hasNext()) {
+        return new Result<>(this.prefixIterator.next().toItem());
+      }
 
-            if (this.error != null) {
-              this.completed = true;
-              return this.error;
-            }
+      this.completed = true;
+      throw new NoSuchElementException();
+    }
 
-            if (this.itemIterator.hasNext()) {
-              Item item = this.itemIterator.next();
-              return new Result<>(item);
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  private Iterable<Result<Item>> listObjectsV2(ListObjectsArgs args) {
+    return new Iterable<Result<Item>>() {
+      @Override
+      public Iterator<Result<Item>> iterator() {
+        return new ObjectIterator() {
+          protected void populateResult()
+              throws InvalidKeyException, InvalidBucketNameException, IllegalArgumentException,
+                  NoSuchAlgorithmException, InsufficientDataException, XmlParserException,
+                  ErrorResponseException, InternalException, InvalidResponseException, IOException {
+            String delimiter = "/";
```

Comment:
We need use `delimiter` from `args.delimiter()` if it is `null` and we can default to `/`

Reply:
I had kept the code as is and just refactored to reduce duplication. However, your suggestion seems like an improvement over the existing code - so I will change the code accordingly.